Alter the plugin to use stricter "to satisfy" semantics.

alexjeffburke opened the pull request:
The behaviour of allowing a set to compare against an array as part
of "to satisfy" has shown itself to be surprising - exclude this
behaviour from "to satisfy" and rename it as a separate assertion
which still allows an array-like on the RHS.

Perhaps more importantly a non-exhaustive satisfy passed with a
subset on the RHS. This is another very surprising behaviour given
arrays in core enforce an equal length. Switch to enforcing the Set
being equal size by default.

papandreou commented on test/unexpected-set.js:
```diff
@@ -165,13 +169,46 @@ describe('unexpected-set', () => {
+            'to satisfy',
+            new Set([{ foo: true }])
+          );
+        });
```
Hmm, didn't we decide to make this case fail?

alexjeffburke replied:
Ah, maybe - my understanding was we decided that much like array-like we'd enforce the size being the same but still honour exhaustively for nested comparisons.

That being said, maybe the fact that sets are percieved to be unordered means this is a bad piece of behaviour. One to ponder some more perhaps.

papandreou replied:
What I took away from our conversation was that this exact case was the main thing to fix for `better-shallow-deep-equal` to DTRT, and that it also made more sense to force `Set` `to satisfy <whatever>` into "exhaustively" mode, similar to `array-like`?

alexjeffburke replied:
The primary issue was allowing `<Set> to satisfy <array-like>` - otherwise we'd end up with comparisons succeeding that we just don't want to. A close second was that "to satisfy" allows subsets by default, which is also an issue.

Once we only allow Set to Set and enforce the same number of items (which is what this PR does so far :) ) then we can separately decide how strict we want the element matching to be. Will write about this separately below.

alexjeffburke commented:
> Ah, nice!
> 
> I thought I volunteered to do this, but luckily I didn't have a chance to work on it yet :)
> 
> Remember to fix the documentation tests too.

Oh I'm sorry if I took your thunder - just thought I'd provide a head start since I had a few spare mins the other day.

papandreou commented:
> Oh I'm sorry if I took your thunder

No worries, I hadn't started :)

alexjeffburke commented on lib/unexpected-set.js:
```diff
@@ -277,23 +277,24 @@ module.exports = {
 
     expect.addAssertion(
-      '<Set> to [exhaustively] satisfy <Set>',
+      '<Set> to have elements satisfying <array-like>',
```
Starting to wonder if this wants to be "members" rather than "elements" on further thought.

papandreou replied:
The assertions for array-like in core are called `to have items satisfying`, would make sense to match that.

alexjeffburke replied:
The reason I'm trying to be different is the RHS of the assertions in core are generally varargs of type <any> - perhaps we should make it vargs and then make the name the same?

papandreou replied:
Hmm, `<array-like> to have items [exhaustively] satisfying` doesn't take varargs? https://github.com/unexpectedjs/unexpected/blob/b3273efd25125310451ba30f409bd4e0a84e6295/lib/assertions.js#L1217

Do you mean the `<assertion>` thing?

alexjeffburke replied:
I’m not sure what I was thinking here - might have been thrown off by the `...rest` but now realise that’s about chained assertions which we probably won’t enable.

Let me make this change & see where we end up.

alexjeffburke commented:
Around the question of whether the members of a set should be allowed to have things missing - the documentation examples might be a good guide.

The thing that gave me pause for thought was the following case:

```output
expected Set([ 1, 2, 3 ]) to satisfy
Set([
  1,
  expect.it('to be less than or equal to', 1),
  expect.it('to be greater than', 10)
])

Set([
  1,
  2, // should be removed
  3 // should be removed
  // missing: should be greater than 10
])
```

The RHS specifies three items, but two of those things are satisfied by the same element on the LHS. I guess what this pokes at is what "exhaustively" really means when applied to Set comparisons.

papandreou commented:
I still think that `<Set> to satisfy <Set>` and `<Set> to [exhaustively] satisfy <Set>` should mean the same thing at the first level. The current semantics are just too surprising.

alexjeffburke commented:
@papandreou turns out there was already a (seemingly) undocumented `"to have items satisfying"` assertion in the plugin - I've adjusted this to the new semantics and added documentation. Please have another look :)

papandreou commented on documentation/assertions/Set/to-have-items-satisfying.md:
````diff
@@ -0,0 +1,57 @@
+Asserts the items contained by a Set satisfy a particular list of items.
+
+```js
+expect(new Set([1, 2, 3]), 'to have items satisfying', [1, { foo: 'bar' }, 3]);
````
This seems wrong. The `to have items satisfying` family of assertions should be used with an assertion (or literal value) that has to hold true for *every item in the collection*, not the collection as a whole. This looks like an alias for `to [exhaustively] satisfy`?

I think it should be used like this:

```js
expect(new Set([1, 2, 3]), 'to have items satisfying', expect.it('to be a number'));
expect(new Set([1]), 'to have items satisfying', 1);
``

alexjeffburke replied:
Yes, you're right - I misread this. Have adjusted the semantics now :)

papandreou commented on documentation/assertions/Set/to-have-items-satisfying.md:
````diff
@@ -0,0 +1,57 @@
+```
+
+In order to check a property holds for all the items, an assertion can be
+passsed as the argument - in this example we assert that all the items in
````
```suggestion
passed as the argument – in this example we assert that all the items in
```

papandreou commented on lib/unexpected-set.js:
```diff
@@ -228,31 +228,46 @@ module.exports = {
 
+    expect.addAssertion('<Set> [not] to be empty', (expect, subject) => {
+      expect(subject.size, '[not] to equal', 0);
+    });
```
:ok_hand:

papandreou commented on lib/unexpected-set.js:
```diff
@@ -228,31 +228,46 @@ module.exports = {
+            }
+          });
+        } else if (valueType.is('array-like')) {
+          expected = nextArg;
```
I think this is the part that's wrong (see above)

papandreou commented on documentation/assertions/Set/to-have-items-satisfying.md:
````diff
@@ -0,0 +1,55 @@
+```
+
+In order to check a property holds for all the items, an assertion can be
+passed as the argument - in this example we assert that all the items in
````
Let's use a dash instead of a hyphen :nerd_face:

```suggestion
passed as the argument – in this example we assert that all the items in
```

papandreou commented on test/unexpected-set.js:
```diff
@@ -163,6 +175,35 @@ describe('unexpected-set', () => {
       }, 'to throw');
     });
+
+    describe('with a chained assertion on the RHS', () => {
```
```suggestion
    describe('with a compound assertion on the RHS', () => {
```
